src/main/scala/nodes/learning/KernelGenerator.scala

```diff
@@ -0,0 +1,273 @@
+package nodes.learning
+
+import breeze.linalg._
+import breeze.math._
+import breeze.numerics._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.SparkConf
+import org.apache.spark.SparkContext
+import org.apache.spark.SparkContext._
+import utils._
+
+
+// Trait to be implemented by Kernel functions
+trait KernelGenerator[T] {
+
+  // TODO(stephentu): these should both be just one function
+
+  /**
+   * Return a column block of the kernel train matrix
+   *
+   * The kernel train matrix is a |train| x |train| symmetric matrix K_train.
+   * This function returns the columns in blockIdxs of K_train  and the block diagonal.
+   */
+  def generateKernelTrainBlock(mat: RDD[T], blockIdxs: Seq[Int]):
+      (RDD[DenseVector[Double]], DenseMatrix[Double])
+
+  /**
+   * Return a column block of the kernel test matrix
+   *
+   * The kernel test matrix is a |test| x |train| matrix K_testtrain.
+   * This function returns the columns in blockIdxs of K_testtrain.
+   *
+   * WARNING: Observe the order that test comes first
+   */
+  def generateKernelTestBlock(
+    testMat: RDD[T],
+    trainMat: RDD[T],
+    blockIdxs: Seq[Int]): RDD[DenseVector[Double]]
+
+}
+
+object SparseLinearKernel extends KernelGenerator[(Array[Long], Array[Double])]
+  with Serializable {
+
+  private def sparseDotProduct(
```

Comment:
Yeah, +1 for leaving it out if we don't have this particular interface figured out. Would be really nice to get in general, though!